### PR TITLE
[ROCm] Fix allreduce + RMSNorm fusion pattern matchin

### PR DESCRIPTION
**Pull request by rbrugaro-amd**
### Summary

Fixes two issues that broke the `allreduce + RMSNorm` fusion pass introduced in #37646, caused by subsequent refactoring in #36823 

1. **`torch.empty_like` → `torch.zeros_like`** in `AiterAllreduceFusedRMSNormPattern._replacement` (`allreduce_rms_fusion.py`):  
   The fused allreduce+rmsnorm kernel always adds `res_inp`; using `empty_like` leaves undefined values that corrupt outputs. Changed to `zeros_like` so the add is a no-op when residual is freshly created.

2. **Conditional `variance_size_override` argument** in `RMSNorm.forward_native` (`layernorm.py`):  
   After the IR refactoring, `ir.ops.rms_norm` and `ir.ops.fused_add_rms_norm.maybe_inplace` were unconditionally passed `self.variance_size_override` (even when `None`). This produced 4-argument calls in the FX graph, but the fusion patterns expect 3 arguments. The mismatch prevented pattern matching entirely. Fixed by conditionally unpacking `variance_size_override` only when it is not `None`.

### Testing

Tested with `Kimi-K2-Thinking-MXFP4` on 4x MI355X (TP=4) 

- **vllm**: `0.20.1rc1.dev153+gcfd2573f2` (base commit `cfd2573f2`)
- **aiter**: `amd-aiter 0.1.12.post2.dev126+g033d8b9db`
- **torch**: `2.10.0+git8514f05`

**Fusion pass results** (confirmed via `VLLM_DEBUG_DUMP_PATH` graph dumps and custom logging):
- `all_reduce_fusion_pass`: **244 pattern matches** across 2 compile ranges (122 per range)
- `mla_dual_rms_norm_fusion_pass`: **183 matches**
- Graph reduced from ~4491 → ~4367 nodes after fusion/cleanup passes
- `fused_add_rms_norm=aiter` implementation selected

### CODE REVIEW

**Comment by attila-dusnoki-htec** [comment]
I tested this with `577b9623e6f8801698d411f4b04269326f5afbe2` base commit + PR#40392 change with kimi-k2.5-mxfp4 model.

*Without* this patch and `"fuse_allreduce_rms": false`:
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9554|±  |0.052|
|     |       |strict-match    |     5|exact_match|↑  |0.9417|±  |0.0067|
```

*Without* this patch and `"fuse_allreduce_rms": true`:
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.008|±  |0.0056|
|     |       |strict-match    |     5|exact_match|↑  |0.004|±  |0.0040|
```

*With* this patch and `"fuse_allreduce_rms": true`:
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.952|±  |0.0135|
|     |       |strict-match    |     5|exact_match|↑  |0.952|±  |0.0135|
```

File: vllm/model_executor/layers/layernorm.py [hunk 1]
```diff
@@ -90,15 +90,23 @@ def forward_native(
                 x,
                 self.weight.data if self.pass_weight else None,
                 self.variance_epsilon,
-                self.variance_size_override,
+                *(
```
Comment (ProExpertProg) [inline vllm/model_executor/layers/layernorm.py:93]: Let's just fix the patterns in the pass instead

Comment (rbrugaro-amd) [reply vllm/model_executor/layers/layernorm.py:93]: Investigated whether the `variance_size_override` argument mismatch between layernorm.py (4 args) and the fusion patterns (3 args) causes a pattern-matching failure. It does not on latest nightly (sha256:18aa8ce6c3f4bf61ae220750311f0f9526bd64e1b419052817accfca625c5f90).

PyTorch's C++ boxed dispatch layer (`parseIValuesToPyArgsKwargs in torch/csrc/autograd/python_variable.cpp`) strips trailing positional arguments that equal their schema defaults. Since variance_size has default=None in the op schema and variance_size_override is None for all standard LLM models, the 4th argument is stripped before it reaches the FX graph the fusion passes operate on. This behavior has been in PyTorch since May 2022 ([55f55a4](https://github.com/pytorch/pytorch/commit/55f55a4cf6cc50cde9e8e8369d92847ca85b23da), PR [#75905](https://github.com/pytorch/pytorch/pull/75905)).

**Graph dump evidence** — tested on vllm/vllm-openai-rocm:nightly (fb1ac806c, PyTorch 2.10.0+git8514f05) with Kimi-K2-Thinking-MXFP4 at TP=4, fresh container with zero variance_size patches:

**__compiled_fn_1.BEFORE_PRE_GRAD.0.py (Dynamo trace)** — rms_norm has 4 args, trailing None from self.variance_size_override:
`rms_norm_default = torch.ops.vllm_ir.rms_norm.default(all_reduce, _get_data_attr, 1e-05, None)`
**__compiled_fn_1.Forward_graph.0.py (AOT Autograd re-trace via C++ dispatch)** — trailing None stripped, 3 args:
`rms_norm = torch.ops.vllm_ir.rms_norm.default(all_reduce, alias, 1e-05)`
**__compiled_fn_1.AFTER_POST_GRAD.0.py (post-fusion)** — correctly replaced with the fused AITER op:

`rocm_aiter_fused_allreduce_rmsnorm_default_121 = torch.ops.vllm.rocm_aiter_fused_allreduce_rmsnorm.default(where, full_default_2, arg3_1, 1e-05)`

245 vllm_ir.rms_norm ops pre-fusion → 366 rocm_aiter_fused_allreduce_rmsnorm ops post-fusion (includes fused_add_rms_norm variants). GSM8K accuracy on Kimi-K2-Thinking-MXFP4: 0.948 (250 samples, 5-shot).


Recommendation: The variance_size_override fix is not needed. The empty_like → zeros_like fix already merged https://github.com/vllm-project/vllm/pull/41972

Does this align with your observation @akii96? if so i will close this PR

**Comment by rbrugaro-amd** [comment]
fix no longer needed
